bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RegionAwareEnsemblePlacementPolicy.java

```diff
@@ -644,12 +645,75 @@ public final DistributionSchedule.WriteSet reorderReadLACSequence(
+
+        int effectiveMinRegionsForDurability = disableDurabilityFeature.isAvailable() ? 1 : minRegionsForDurability;
+
+        int ensembleSize = ensembleList.size();
+        Map<String, Set<BookieId>> regionsInQuorum = new HashMap<>();
+        BookieId bookie;
+        for (int i = 0; i < ensembleList.size(); i++) {
+            regionsInQuorum.clear();
+            for (int j = 0; j < writeQuorumSize; j++) {
+                bookie = ensembleList.get((i + j) % ensembleSize);
+                if (knownBookies.containsKey(bookie)) {
+                    String region = getLocalRegion(knownBookies.get(bookie));
+                    if (regionsInQuorum.containsKey(region)) {
+                        regionsInQuorum.get(region).add(bookie);
+                    } else {
+                        Set<BookieId> bookieSet = new HashSet<>();
+                        bookieSet.add(bookie);
+                        regionsInQuorum.put(region, bookieSet);
+                    }
+                } else if (LOG.isDebugEnabled()) {
+                    LOG.debug("bookie {} is not in the list of knownBookies", bookie);
+                }
+            }
+
+            if (regionsInQuorum.isEmpty()) {
+                return PlacementPolicyAdherence.FAIL;
+            }
+
+            if (regionsInQuorum.size() < 2) {
+                // fall back to use the ensemblePlacementPolicy in specific region
+                String region = regionsInQuorum.keySet().iterator().next();
+                Set<BookieId> bookieIds = regionsInQuorum.get(region);
+
+                TopologyAwareEnsemblePlacementPolicy policyWithinRegion = perRegionPlacement.get(region);
+                PlacementPolicyAdherence isEnsembleAdheringToPlacementPolicy = policyWithinRegion
+                        .isEnsembleAdheringToPlacementPolicy(new ArrayList<>(bookieIds), bookieIds.size(), 1);
+                if (isEnsembleAdheringToPlacementPolicy == PlacementPolicyAdherence.FAIL) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("For ensemble {}, write set starting at {} are all from one region, "
+                                + "fall back to RackawareEnsemblePlacementPolicy and fail.", ensembleList, i);
+                    }
+                    return PlacementPolicyAdherence.FAIL;
+                }
+                continue;
+            }
+
+            if (effectiveMinRegionsForDurability > 0 && regionsInQuorum.size() < effectiveMinRegionsForDurability) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("For ensemble {}, write set starting at {} are from {} regions, "
+                                    + "less than effectiveMinRegionsForDurability: {}.",
+                            ensembleList, i, regionsInQuorum.size(), effectiveMinRegionsForDurability);
+                }
+                return PlacementPolicyAdherence.FAIL;
+            }
+
+            if (regionsInQuorum.size() < writeQuorumSize) {
```

Comment:
We can check it before line_679

Reply:
We should judge `regionsInQuorum.size() < 2` first, in case that it falls back to `RackawareEnsemblePlacementPolicy`.